Api v3

escopecz opened the pull request:
https://mautic.atlassian.net/browse/TPROD-346

These changes are needed for https://github.com/mautic/mautic/pull/11724

This PR removes endpoints that the Mautic project do not use as we have no resources to maintain all the endpoints. Support for other endpoints can be added any time. As this is a major release and can contain breaking changes I decided to go with this approach as the API v2 will stop working in few days.

I was developing with `tests/Functional/LiveTest.php` which makes all the queries we'll need to make sure it's working correctly. Then there are bunch of unit tests that can run in the CI.

This PR is also moving the scattered docs into `Readme.md` with useful examples.

The changes here can be tested with a live Transifex project where you are a maintainer and this project has a `cs` translation. DO NOT TEST WITH A PRODUCTION PROJECT. More details in the [readme.md](https://github.com/mautic/Transifex-API/pull/4/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)

escopecz committed:
Updating min PHP from 7.2 to 7.4 and dev packages

escopecz committed:
Removing unused property

escopecz committed:
Updating CS fixer config to new version

escopecz committed:
Adding PHP 8.1 to the supported versions

RCheesley commented:
Is there any way to test this? Guessing maybe not?

escopecz commented:
It's not yet done. I'm currently discussing whit Transifex support why one endpoint works for me and another gives me 403 permission denied response. So that's currently a blocker but it will be quite a refactoring. It will be a BC break obviously so I also updated dependencies and PHP versions starting with 7.4 so we could support M4 too.

RCheesley commented:
Ah OK! Good luck with getting to the bottom of it all!

I guess if it's a BC break we have to wait for Mautic 5 - does this mean that languages will potentially stop working for Mautic 4 until we get that released?

escopecz commented:
I meant that it will be a BC break for this repository/library so we'll have to release it as a major version. We can implement it for M4. It would break some plugins using this library but they would stop working as the old API is getting discontinued anyway so upgrading it to a working API is not as bad.

RCheesley commented:
Ah OK that makes sense, thanks for the clarification!

escopecz committed:
Introducing new config class to replace the existing "options" array

escopecz committed:
We'll be using FQCN instead of loading files and api_key instead of username and password

escopecz committed:
Updating ApiFactoryTest to use config

escopecz committed:
Updating docs to use FQCN over strings and new API reference links

escopecz committed:
Adding link to the Changelog to where the new versions are

escopecz committed:
Using the exception that tests are written for to keep at least some BC

escopecz committed:
Validate config if the connector is used without the factory

escopecz committed:
Updating some endpoints to the new API

escopecz committed:
Updating tests to use config over options

escopecz committed:
Pulling language translations ready

escopecz committed:
Removing endpoints that won't make it to the new major version

as the Mautic project do not use those and is lacking resources to support code that is unused

escopecz committed:
Using new phpunit config format, adding env vars for functional tests

escopecz committed:
Removing ApiFactory as the same can be done simply in the Transifex class

escopecz committed:
Allowing Config to be configured from env vars

escopecz committed:
Improving error reporting for faster debugging

escopecz committed:
Finishing Resources endpoints and covering them with functional tests

escopecz committed:
Simplifying the work with promises

escopecz committed:
Finishing functional test

escopecz committed:
Starting to move docs to the readme only

escopecz committed:
Updating unit tests

escopecz committed:
Moving docs from separate dir to readme

escopecz committed:
Removig unused exceptions

escopecz committed:
Setting PHP to the minimal supported version so dependencies with higher version are not installed

escopecz committed:
CS fixes

escopecz committed:
Skipping the live test for CI checks

escopecz committed:
Fixing STAN

escopecz committed:
Making the method name more concrete as there are also resource stats

sdoering reviewed (commented):
Tested the PR in Gitpod by following the helpful hints from the documentation and by @escopecz . Thanks for helping me doing my first PR testing (ever) on an OS project. 

Steps I followed: 

1. Open the PR in gitpod
2. composer install dependencies
3. run `composer test` with the following result: 

```
gitpod /workspace/Transifex-API (api-v3) $ composer test
> vendor/bin/phpunit
PHPUnit 9.5.26 by Sebastian Bergmann and contributors.

.....................                                             21 / 21 (100%)

Time: 00:04.526, Memory: 6.00 MB

OK (21 tests, 94 assertions)
```
4. Did separate steps for functional and unittests, just to be sure: 
```
gitpod /workspace/Transifex-API (api-v3) $ composer test -- --testsuite=Functional
> vendor/bin/phpunit '--testsuite=Functional'
PHPUnit 9.5.26 by Sebastian Bergmann and contributors.

.                                                                   1 / 1 (100%)

Time: 00:04.539, Memory: 6.00 MB

OK (1 test, 32 assertions)
```
```
gitpod /workspace/Transifex-API (api-v3) $ composer test -- --testsuite=Unit
> vendor/bin/phpunit '--testsuite=Unit'
PHPUnit 9.5.26 by Sebastian Bergmann and contributors.

....................                                              20 / 20 (100%)

Time: 00:00.009, Memory: 6.00 MB

OK (20 tests, 62 assertions)
```
5. Ran codestyle checks: 
```
gitpod /workspace/Transifex-API (api-v3) $ composer cs
> vendor/bin/php-cs-fixer fix -v --dry-run --diff
PHP CS Fixer 3.13.0 Oliva by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.1.12
Loaded config default from "/workspace/Transifex-API/.php-cs-fixer.php".
Using cache file ".php-cs-fixer.cache".
SSSSSSSSSSSSSSSSSSSSSSSSSS                                                                                                                             26 / 26 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error

Checked all files in 0.003 seconds, 14.000 MB memory used

Detected deprecations in use:
- Rule "no_trailing_comma_in_singleline_array" is deprecated. Use "no_trailing_comma_in_singleline" instead.
```
6. Ran static analysis checks:
```
gitpod /workspace/Transifex-API (api-v3) $ composer phpstan
> vendor/bin/phpstan analyse src tests -l 5
 26/26 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] No errors                                                                                                         
                                                                                                                        
```

All checks ran 100% fine - thanks again for enabling me to support Mautic. 

Best regards, 
Sven

tejasync commented on README.md:
````diff
@@ -0,0 +1,236 @@
+);
+```
+
+#### Download a traslation
````
```suggestion
#### Download a translation
```
Nitpicking.

escopecz committed:
Typo fix

Co-authored-by: Tejas Navghane <ts.navghane@gmail.com>

tejasync reviewed (approved):
The code look good to me! 🎉  
Nicely done, thank you @escopecz!👌🏻
